Adobe-Consulting-Services/acs-aem-commons
# New Feature: Sitewide Component Properties

This is an **alpha** state feature (not yet vetted for production use) that HS2 Solutions put together at a company-wide 1 & 1/2 day hackathon.  We will have a PR for the documentation page as well, but wanted to get this up to see if people think this is a reasonable feature to add, or if AEM has a better way of doing this that we just couldn't find.

**Problem being solved:**
- Multiple instances of the same AEM component need to be configured separately, leading to:
  - Excessive copy/paste (error-prone)
  - No reuse of properties across instances without custom implementation
  - Labor-intensive and unmanageable for content changes
- Reference component solves some cases, but
  - Is an all or nothing deal - the reference is an exact copy of the parent
  - If not used carefully, Reference components can create a nasty spiderweb of content references across the site

**Goals of our solution:**
- No alteration of existing functionality - instance level component properties are unaffected
- New site-wide properties can be added to ensure consistency of a component's content across the entire site.  With site-wide properties, an update covers the entire site for that component (no copy and paste).
- Support overriding site-side properties by instance-level properties when set.
- Site-wide properties dialog has parity with OOTB instance-level properties dialog

**Installation:**
This feature is enabled by adding a cq:infoProvider to your page component, referencing com.adobe.acs.commons.wcm.impl.SitewideComponentPropsPageInfoProvider. See documentation at https://docs.adobe.com/docs/en/aem/6-1/develop/components/pageinfo.html for adding nodes to cq:infoProvider.

See our example implementation component at /apps/acs-commons/components/content/sitewide-component-props-example - the magic is added by the cq:dialogsitewide node which mimics the standard cq:dialog node.

**Still need to test/verify:**
- Ensure that properties can be sent out for translation
- Ensure that Reference component can work with global configs
- Ensure that path fields within global configs are updated if referenced paths are updated
- Ensure that merging of dialogsitewide.xml from resourceSuperType is working

**Future Features and Enhancements:**
- Only display the sitewide properties button for components that define a _cq_dialogsitewide node — Alternatively show a "no sitewide configs" message for components that do not have them
- Move the sitewide properties button on the authoring toolbar next to the wrench icon for the standard properties dialog
- Ensure authoring page refreshes when sitewide props are updated, since multiple component instances may be affected
- Flush dispatcher for all pages within the site using a component when site-wide properties are updated.
- Add support for site-wide properties that are not specific to a component type
